File: app/src/main/java/com/jodelXposed/krokofant/utils/Bitmap.java
package com.jodelXposed.krokofant.utils;

import android.graphics.*;
import android.os.Environment;

import java.io.*;

import static com.jodelXposed.krokofant.utils.Log.xlog;

public final class Bitmap {
    public static String jodelImagePath = Environment.getExternalStorageDirectory() + "/.jodel-input.jpg";

    public static android.graphics.Bitmap loadBitmap() {
        xlog("Loading bitmap image");
        File file = new File(jodelImagePath);
        return file.exists() ? bitmapFromFile(file) : bitmapFromFile(new File(jodelImagePath));
    }

    public static void saveBitmap(android.graphics.Bitmap bitmap) {
        xlog("Saving bitmap of size: " + bitmap.getByteCount());

        try {
            FileOutputStream fos = new FileOutputStream(new File(jodelImagePath));
            bitmap.compress(android.graphics.Bitmap.CompressFormat.JPEG, 100, fos);
            fos.close();
        } catch (FileNotFoundException e) {
            xlog("File not found: " + e.getLocalizedMessage());
        } catch (IOException e) {
            xlog("Error accessing file: " + e.getLocalizedMessage());
        }
    }

    public static android.graphics.Bitmap bitmapFromFile(File file) {
        ByteArrayOutputStream blob = new ByteArrayOutputStream();
        android.graphics.Bitmap bitmap = BitmapFactory.decodeFile(file.getAbsolutePath());
        if(bitmap != null) bitmap.compress(android.graphics.Bitmap.CompressFormat.JPEG, 100, blob);
        else xlog("Loaded bitmap is null");
        return bitmap;
    }

    public static android.graphics.Bitmap toGrayscale(android.graphics.Bitmap bmpOriginal) {
        int width, height;
        height = bmpOriginal.getHeight();
        width = bmpOriginal.getWidth();

        android.graphics.Bitmap bmpGrayscale = android.graphics.Bitmap.createBitmap(width, height, android.graphics.Bitmap.Config.ARGB_8888);
        Canvas c = new Canvas(bmpGrayscale);
        Paint paint = new Paint();
        ColorMatrix cm = new ColorMatrix();
        cm.setSaturation(0);
        ColorMatrixColorFilter f = new ColorMatrixColorFilter(cm);
        paint.setColorFilter(f);
        c.drawBitmap(bmpOriginal, 0, 0, paint);
        return bmpGrayscale;
    }
}

File: app/src/main/java/com/jodelXposed/utils/Bitmap.kt
package com.jodelXposed.utils

import android.graphics.BitmapFactory
import android.media.MediaScannerConnection
import com.jodelXposed.utils.Log.dlog
import com.jodelXposed.utils.Log.xlog
import com.jodelXposed.utils.Utils.getJXSharedImage
import java.io.*


object Bitmap {
    @JvmStatic
    fun loadBitmap(): android.graphics.Bitmap {
        dlog("Loading bitmap image")
        return BitmapFactory.decodeFile(getJXSharedImage())
    }

    @JvmStatic
    fun loadBitmap(inputStream: InputStream): android.graphics.Bitmap {
        dlog("Loading bitmap image")
        val byteArray: ByteArray = inputStream.readBytes()
        inputStream.close()
        return BitmapFactory.decodeByteArray(byteArray, 0, byteArray.size)
    }

    @JvmStatic
    fun saveBitmap(bitmap: android.graphics.Bitmap) {
        saveBitmap(bitmap, null)
    }

    @JvmStatic
    fun saveBitmap(bitmap: android.graphics.Bitmap, path: String?) {
        try {
            dlog("Saving bitmap of size: ${bitmap.byteCount}")
            var fos: FileOutputStream? = null
            val file: File?
            if (path == null) {
                file = File(getJXSharedImage())
                if (!file.exists())
                    file.createNewFile()
                fos = FileOutputStream(file)
            } else {
                file = File(path)
                if (!file.exists())
                    file.createNewFile()
                fos = FileOutputStream(file)
            }
            bitmap.compress(android.graphics.Bitmap.CompressFormat.JPEG, 100, fos)
            fos.close()

            scanFile(file)

        } catch (e: FileNotFoundException) {
            xlog("File not found", e)
        } catch (e: IOException) {
            xlog("Error accessing file", e)
        }

    }

    fun scanFile(file: File){
        try {
            MediaScannerConnection.scanFile(XposedUtilHelpers.getActivityFromActivityThread(),
                    arrayOf(file.toString()), null
            ) { path, uri ->
                Log.dlog("ExternalStorageScanned $path:")
                Log.dlog("ExternalStorage" + "-> uri=" + uri)
            }
        }catch (e :Exception){
            e.printStackTrace()
        }
    }

}